src/main/scala/mesosphere/marathon/api/LeaderProxyFilter.scala

```diff
@@ -1,137 +1,263 @@
+      log.error("Prevent proxy cycle, rejecting request")
+      response.sendError(HttpStatus.SC_BAD_GATEWAY, ERROR_STATUS_LOOP)
+      return
+    }
+
+    val method = request.getMethod
+
+    log.info(s"Proxying request to $method $url from $myHostPort")
+
+    val leaderConnection: HttpURLConnection = url.openConnection().asInstanceOf[HttpURLConnection]
+
+    leaderConnection.setConnectTimeout(leaderProxyConf.leaderProxyConnectionTimeout())
+    leaderConnection.setReadTimeout(leaderProxyConf.leaderProxyReadTimeout())
+
+    // getHeaderNames() and getHeaders() are known to return null, see:
+    //http://docs.oracle.com/javaee/6/api/javax/servlet/http/HttpServletRequest.html#getHeaders(java.lang.String)
+    val names = Option(request.getHeaderNames).map(_.asScala).getOrElse(Nil)
+    for {
+      name <- names if !name.equalsIgnoreCase("host") && !name.equalsIgnoreCase("connection")
+      headerValues <- Option(request.getHeaders(name))
+      headerValue <- headerValues.asScala
+    } {
+      log.debug(s"addRequestProperty $name: $headerValue")
+      leaderConnection.addRequestProperty(name, headerValue)
+    }
+
+    leaderConnection.addRequestProperty(HEADER_VIA, viaValue)
 
+    leaderConnection.setRequestMethod(method)
+
+    log.debug("request properties {}", leaderConnection.getRequestProperties.asScala)
+
+    method match {
+      case "GET" | "HEAD" | "DELETE" =>
+        leaderConnection.setDoOutput(false)
+      case _ =>
+        leaderConnection.setDoOutput(true)
+
+        val proxyOutputStream = leaderConnection.getOutputStream
+        try {
+          copy(request.getInputStream, proxyOutputStream)
+        }
+        finally {
+          Try(proxyOutputStream.close())
         }
-        catch {
-          case e: Exception =>
-            log.warn("Exception while proxying", e)
+    }
+
+    try {
+      val status = leaderConnection.getResponseCode
+      response.setStatus(status)
+
+      val fields = leaderConnection.getHeaderFields
+      // getHeaderNames() and getHeaders() are known to return null
+      if (fields != null) {
+        for ((name, values) <- fields.asScala) {
+          if (name != null && values != null) {
+            for (value <- values.asScala) {
+              response.addHeader(name, value)
+            }
+          }
         }
       }
+      copy(leaderConnection.getInputStream, response.getOutputStream)
+    }
+    catch {
+      case connException: ConnectException =>
+        response.sendError(HttpStatus.SC_BAD_GATEWAY, ERROR_STATUS_CONNECTION_REFUSED)
+      case NonFatal(e) =>
+        copy(leaderConnection.getErrorStream, response.getOutputStream)
+    }
+    finally {
+      Try(response.getOutputStream.close())
+      Try(leaderConnection.getInputStream.close())
+      Try(leaderConnection.getErrorStream.close())
     }
   }
 
-  def destroy() {
-    //NO-OP
+  private[this] def copy(inputOrNull: InputStream, outputOrNull: OutputStream): Unit = {
```

Comment:
There is an IO trait with such functionality. Please use/adapt existing functionality.